[feat] IP Adapters (author @okotaku ) (#5713)
huggingface/diffusers: docs/source/en/using-diffusers/loading_adapters.md

image = pipeline(prompt=prompt).images[0]
image
```

### IP-Adapter 

[IP-Adapter](https://ip-adapter.github.io/) is an effective and lightweight adapter to achieve image prompt capability for the pre-trained text-to-image diffusion models. It is now available to use with most of our Stable Diffusion and Stable Diffusion XL pipelines. You can also use the IP-Adapter with other custom models fine-tuned from the same base model, as well as ControlNet and T2I adapters. Moreover, the image prompt can also work well with the text prompt to accomplish multimodal image generation.

You can find the officially available IP-Adapter checkpoints in [h94/IP-Adapter](https://huggingface.co/h94/IP-Adapter).

IP-Adapter was contributed by [okotaku](https://github.com/okotaku).

Let's look at an example where we use IP-Adapter with the Stable Diffusion text-to-image pipeline. 

> ``` py

Reviewer: ```suggestion
```py
```

from diffusers import AutoPipelineForText2Image, CLIPVisionModelWithProjection
import torch
from diffusers.utils import load_image

image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", 
    subfolder="models/image_encoder",
    torch_dtype=torch.float16,
).to("cuda")

pipeline = AutoPipelineForText2Image.from_pretrained("runwayml/stable-diffusion-v1-5", image_encoder=image_encoder, torch_dtype=torch.float16).to("cuda")
```

> Now you can load the IP-Adapter with [`~loaders.IPAdapterMixin.load_ip_adapter`] method. 

Reviewer 2: ```suggestion
Now load the [h94/IP-Adapter](https://huggingface.co/h94/IP-Adapter) weights with the [`~loaders.IPAdapterMixin.load_ip_adapter`] method. 
```

```py
pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter_sd15.bin")
```

IP-Adapter allows you to use both image and text to condition the image generation process. In this example, let's take the cute bear eating pizza that we generated with Textual Inversion, and create a new bear that is similarly cute but wears sunglasses. We can pass the bear image as `ip_adapter_image`, along with a text prompt that mentions "sunglasses". 

```py
pipeline.set_ip_adapter_scale(0.6)
image = load_image("https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/load_neg_embed.png")
generator = torch.Generator(device="cpu").manual_seed(33)
images = pipeline(
    prompt='best quality, high quality, wearing sunglasses', 
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=50,
    generator=generator,
).images
images[0]
```

<div class="flex justify-center">
    <img src="https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/ip-bear.png" />
</div>

<Tip>

> You can use the `pipeline.set_ip_adapter_scale()` method to adjust the ratio of text prompt and image prompt condition.  If you only use the image prompt, you should set the scale to be `1.0`. You can lower the scale to get more diversity in the generation, at the cost of less prompt alignment.

Reviewer 2: ```suggestion
You can use the [`~loaders.IPAdapterMixin.set_ip_adapter_scale`] method to adjust the text prompt and image prompt condition ratio.  If you're only using the image prompt, you should set the scale to `1.0`. You can lower the scale to get more generation diversity, but it'll be less aligned with the prompt.
```

`scale=0.5` can achieve good results in most cases when you use both text and image prompts.
</Tip>

> IP-Adapter also works great with Image-to-Image and Inpainting pipelines. Here is an example of how you can use it with Image-to-Image.

Reviewer 2: Since most of the code is pretty similar for the image-to-image example, what do you think about using it with the new `hfoptions` syntax to make it a bit cleaner (you can even throw in the inpaint example if you want)? For example:

```md
<hfoptions id="tasks">
<hfoption id="text-to-image">

text + code example

</hfoption>
<hfoption id="image-to-image">

text + code example

</hfoption>
</hfoptions>
```

It'll look like:
![Screenshot 2023-11-09 at 9 12 52 AM](https://github.com/huggingface/diffusers/assets/59462357/79230461-785a-4ee0-a358-6f33577ad3d5)

```py
from diffusers import AutoPipelineForImage2Image
import torch
from diffusers.utils import load_image

image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", 
    subfolder="models/image_encoder",
    torch_dtype=torch.float16,
).to("cuda")

pipeline = AutoPipelineForImage2Image.from_pretrained("runwayml/stable-diffusion-v1-5", image_encoder=image_encoder, torch_dtype=torch.float16).to("cuda")

image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/vermeer.jpg")
ip_image = load_image("https://huggingface.co/datasets/YiYiXu/testing-images/resolve/main/river.png")

pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="models", weight_name="ip-adapter_sd15.bin")
generator = torch.Generator(device="cpu").manual_seed(33)
images = pipeline(
    prompt='best quality, high quality', 
    image = image,
    ip_adapter_image=ip_image,
    num_inference_steps=50,
    generator=generator,
    strength=0.6,
).images
images[0]
```

> IP-Adapters can be used with [Stable Diffusion XL](../api/pipelines/stable_diffusion/stable_diffusion_xl.md) (SDXL) for text-to-image, image-to-image, and inpainting pipelines. Below is an example for SDXL text-to-image.

Reviewer 2: I think this way it keeps the focus more on loading

```suggestion
IP-Adapters can also be used with [SDXL](../api/pipelines/stable_diffusion/stable_diffusion_xl.md), but you'll also need to load a [`~transformers.CLIPImageProcessor`] as your feature extractor and pass it to the pipeline.
```

from diffusers.utils import load_image
from transformers import CLIPVisionModelWithProjection, CLIPImageProcessor
import torch

image_encoder = CLIPVisionModelWithProjection.from_pretrained(
    "h94/IP-Adapter", 
    subfolder="sdxl_models/image_encoder",
    torch_dtype=torch.float16,
).to("cuda")
feature_extractor = CLIPImageProcessor.from_pretrained("laion/CLIP-ViT-bigG-14-laion2B-39B-b160k")
  
pipeline = AutoPipelineForText2Image.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0",
    image_encoder=image_encoder,
    feature_extractor=feature_extractor,
    torch_dtype=torch.float16
).to("cuda")

image = load_image("https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/watercolor_painting.jpeg")

pipeline.load_ip_adapter("h94/IP-Adapter", subfolder="sdxl_models", weight_name="ip-adapter_sdxl.bin")

generator = torch.Generator(device="cpu").manual_seed(33)
image = pipeline(
    prompt="best quality, high quality", 
    ip_adapter_image=image,
    negative_prompt="monochrome, lowres, bad anatomy, worst quality, low quality", 
    num_inference_steps=25,
    generator=generator,
).images[0]
image.save("sdxl_t2i.png")
```

> <div class="flex justify-center">
>    <table border="1">
>     <tr>
>         <th>Input Image</th>
>         <th>Adapted Image</th>
>     </tr>
>     <tr>
>         <td><img src="https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/watercolor_painting.jpeg" alt="Input Image"></td>
>         <td><img src="https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/sdxl_t2i.png" alt="Adapted Image"></td>
>     </tr>
>     </table>
> </div>

Reviewer 2: I think it'll look nicer like this without the line at the top

```suggestion
<div class="flex flex-row gap-4">
  <div class="flex-1">
    <img class="rounded-xl" src="https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/watercolor_painting.jpeg"/>
    <figcaption class="mt-2 text-center text-sm text-gray-500">input image</figcaption>
  </div>
  <div class="flex-1">
    <img class="rounded-xl" src="https://huggingface.co/datasets/sayakpaul/sample-datasets/resolve/main/sdxl_t2i.png"/>
    <figcaption class="mt-2 text-center text-sm text-gray-500">adapted image</figcaption>
  </div>
</div>
```

image = pipeline(prompt=prompt).images[0]
image
```

### IP-Adapter 

[IP-Adapter](https://ip-adapter.github.io/) is an effective and lightweight adapter to achieve image prompt capability for the pre-trained text-to-image diffusion models. It is now available to use with most of our Stable Diffusion and Stable Diffusion XL pipelines. You can also use the IP-Adapter with other custom models fine-tuned from the same base model, as well as ControlNet and T2I adapters. Moreover, the image prompt can also work well with the text prompt to accomplish multimodal image generation.

Let's look at an example where we use IP-Adapter with the Stable Diffusion text-to-image pipeline.

``` py
from diffusers import AutoPipelineForText2Image
import torch
from diffusers.utils import load_image

> pipeline = AutoPipelineForText2Image.from_pretrained("runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16).to("cuda")

Reviewer: This doesn't load the `image_encoder`.

Author: oh yeah we are not going to add it to the sd folder?

Author: we can add it right? it is pretrained

Reviewer: Discussed via Slack. Since the original checkpoints don't ship the `image_encoder` officially, it won't make sense for us to add it now I think. So, it's better to be explicit here and it doesn't hurt either as it's just a one-liner to initialize the `image_encoder`.

Reviewer 3: No we won't be able to add this to the SD folder. We'll have to always pass it manually
